spring-security-modules/spring-security-web-boot-4/src/main/java/com/baeldung/apikeysecretauth/security/ApiKeySecretAuthenticationProvider.java

```diff
@@ -0,0 +1,51 @@
+package com.baeldung.apikeysecretauth.security;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.authentication.BadCredentialsException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+import com.baeldung.apikeysecretauth.model.User;
+import com.baeldung.apikeysecretauth.repository.UserKeysRepository;
+import com.baeldung.apikeysecretauth.repository.model.UserKeysData;
+
+public class ApiKeySecretAuthenticationProvider implements AuthenticationProvider {
+    private final PasswordEncoder passwordEncoder;
+    private final UserKeysRepository userKeysRepository;
+
+    public ApiKeySecretAuthenticationProvider(PasswordEncoder passwordEncoder, UserKeysRepository userKeysRepository) {
+        this.passwordEncoder = passwordEncoder;
+        this.userKeysRepository = userKeysRepository;
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        final ApiKeySecretAuthenticationToken token = (ApiKeySecretAuthenticationToken) authentication;
+        final UserKeysData userKey = userKeysRepository.findById(token.getApiKey())
```

Comment:
Again, let's remove the `final` statements here. Though there are coding standards which prefer this approach, there's no _need_ for it, and it adds noise to the code.

Reply:
fixed